### PR TITLE
tests for newChain, write

### PR DESCRIPTION
**Pull request by cranium7811**
- redeem()
- testNewChain()
- testFuzzNewChain()
- testWrite()
- testFuzzWrite()
- testExercise()

*Changed _nextTokenId and chainMap to public for testing

### CODE REVIEW

File: src/test/OptionSettlement.t.sol [hunk 1]
```diff
@@ -81,26 +91,112 @@ contract OptionSettlementTest is DSTest, NFTreceiver {
...
 
-    function testNewChain(uint256 settlementSeed) public {
+    function testNewChain() public {
+        uint256 nextTokenId = engine._nextTokenId();
+
         Option memory info = Option({
             underlyingAsset: address(weth),
             exerciseAsset: address(dai),
-            settlementSeed: settlementSeed,
+            settlementSeed: 0,
```
Comment (0xAlcibiades) [inline src/test/OptionSettlement.t.sol:114]: This explicitly should not be 0 for the purpose of the test. I want to make sure it gets cleared in the function, whatever the value is.

File: src/OptionSettlement.sol [hunk 2]
```diff
@@ -70,7 +71,7 @@ contract OptionSettlementEngine is ERC1155 {
     address public feeTo = 0x36273803306a3C22bc848f8Db761e974697ece0d;
 
     // To increment the next available token id
-    uint256 private _nextTokenId;
+    uint256 public _nextTokenId;
```
Comment (0xAlcibiades) [inline src/OptionSettlement.sol:74]: If we want this to be public then we should remove the leading underscore. I'm not sure about making it public.

**Review by 0xAlcibiades — CHANGES_REQUESTED** [review]
A few changes requested, but overall really awesome tests.

File: src/OptionSettlement.sol [hunk 3]
```diff
@@ -264,9 +265,43 @@ contract OptionSettlementEngine is ERC1155 {
...
+        uint256 optionId = claim[claimId].option;
+
+        Option storage optionRecord = option[optionId];
+
+        require(
+            optionRecord.expiryTimestamp <= block.timestamp,
+            "Not expired yet"
+        );
+
+        uint256 tx_amount = optionRecord.underlyingAmount *
```
[the remark this answers is not in the document]
Comment (0xAlcibiades) [reply src/OptionSettlement.sol:284]: Let's pull this entire code block out to a separate PR/branch and keep this one about test instrumentation.

File: foundry.toml [hunk 4]
```diff
@@ -2,6 +2,5 @@
 src = 'src'
 out = 'out'
 libs = ['lib']
-remappings = ['ds-test/=lib/ds-test/src/']
```
[the remark this answers is not in the document]
Comment (cranium7811) [reply foundry.toml:5]: changed.

File: remappings.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+base64/=lib/base64/src/
```
[the remark this answers is not in the document]
Comment (cranium7811) [reply remappings.txt]: updated.

File: src/test/OptionSettlement.t.sol [hunk 6]
```diff
@@ -45,12 +45,21 @@ contract OptionSettlementTest is DSTest, NFTreceiver {
...
+    uint256 public wethTotalSupply;
+    uint256 public daiTotalSupply;
+
     IWETH public weth;
     IERC20 public dai;
-    OptionSettlementEngine public engine;
 
     using stdStorage for StdStorage;
-    StdStorage stdstore;
+    StdStorage public stdstore;
```
[the remark this answers is not in the document]
Comment (cranium7811) [reply src/test/OptionSettlement.t.sol]: had a lint warning. updated.

File: src/test/OptionSettlement.t.sol (hunk 1 above)
Comment (cranium7811) [reply src/test/OptionSettlement.t.sol:109]: that's why i'm writing two tests for each method. one will hard coded values and the other with fuzzing.

File: src/OptionSettlement.sol (hunk 2 above)
Comment (cranium7811) [reply src/OptionSettlement.sol:73]: as mentioned above, changed _nextTokenId and chainMap to public just for testing. maybe we can change it to private at the end?

File: src/test/OptionSettlement.t.sol [hunk 7]
```diff
@@ -49,6 +49,7 @@ contract OptionSettlementTest is DSTest, NFTreceiver {
 
     Type public testOptionType = Type.Option;
```
[the remark this answers is not in the document]
Comment (cranium7811) [reply src/test/OptionSettlement.t.sol]: updated.

File: src/test/OptionSettlement.t.sol (hunk 1 above)
Comment (0xAlcibiades) [reply src/test/OptionSettlement.t.sol:109]: Cool

File: src/OptionSettlement.sol (hunk 2 above)
Comment (0xAlcibiades) [reply src/OptionSettlement.sol:73]: Ok, then let's just remove the leading underscore here, which denotes a private storage variable.

Comment (cranium7811) [reply src/OptionSettlement.sol]: done.

File: src/OptionSettlement.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (cranium7811) [reply src/OptionSettlement.sol]: created a new PR for feature/redeem.

File: src/OptionSettlement.sol [hunk 8]
```diff
@@ -264,9 +265,43 @@ contract OptionSettlementEngine is ERC1155 {
...
+
+        tokenType[claimId] = Type.None;
+        claim[claimId].claimed = true;
+
+        uint256[] memory tokens = new uint256[](2);
+        tokens[0] = optionId;
+        tokens[1] = claimId;
+
+        uint256[] memory amounts = new uint256[](2);
+        amounts[0] = claim[claimId].amountWritten;
```
[the remark this answers is not in the document]
Comment (cranium7811) [reply src/OptionSettlement.sol]: created a new PR for feature/redeem.

**Review by 0xAlcibiades — APPROVED** [review]

**Pull request merged**